pkg/controllers/networkpolicy/networkpolicy_controller.go

```diff
@@ -0,0 +1,213 @@
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, MetricResType)
+}
+
+func k8sClient(mgr ctrl.Manager) client.Client {
+	var c client.Client
+	if mgr != nil {
+		c = mgr.GetClient()
+	}
+	return c
+}
+
+func deleteFail(r *NetworkPolicyReconciler, c *context.Context, o *networkingv1.NetworkPolicy, e *error) {
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, MetricResType)
+}
+
+func updateSuccess(r *NetworkPolicyReconciler, c *context.Context, o *networkingv1.NetworkPolicy) {
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateSuccessTotal, MetricResType)
+}
+
+func deleteSuccess(r *NetworkPolicyReconciler, _ *context.Context, _ *networkingv1.NetworkPolicy) {
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, MetricResType)
+}
+
+func (r *NetworkPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	networkPolicy := &networkingv1.NetworkPolicy{}
+	log.Info("reconciling networkpolicy", "networkpolicy", req.NamespacedName)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, MetricResType)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, networkPolicy); err != nil {
+		log.Error(err, "unable to fetch network policy", "req", req.NamespacedName)
+		return ResultNormal, client.IgnoreNotFound(err)
+	}
+
+	// Since NetworkPolicy service can only be activated from NSX 3.2.0 onwards,
+	// So need to check NSX version before starting NetworkPolicy reconcile
+	if !r.Service.NSXClient.NSXCheckVersion(nsx.SecurityPolicy) {
+		err := errors.New("NSX version check failed, SecurityPolicy feature is not supported")
+		updateFail(r, &ctx, networkPolicy, &err)
+		// if NSX version check fails, it will be put back to reconcile queue and be reconciled after 5 minutes
+		return ResultRequeueAfter5mins, nil
+	}
+
+	if networkPolicy.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, MetricResType)
+		if !controllerutil.ContainsFinalizer(networkPolicy, servicecommon.NetworkPolicyFinalizerName) {
+			controllerutil.AddFinalizer(networkPolicy, servicecommon.NetworkPolicyFinalizerName)
+			if err := r.Client.Update(ctx, networkPolicy); err != nil {
+				log.Error(err, "add finalizer", "networkpolicy", req.NamespacedName)
+				updateFail(r, &ctx, networkPolicy, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("added finalizer on networkpolicy", "networkpolicy", req.NamespacedName)
+		}
+
+		if isCRInSysNs, err := util.IsSystemNamespace(r.Client, req.Namespace, nil); err != nil {
+			err = errors.New("fetch namespace associated with network policy failed")
+			log.Error(err, "would retry exponentially", "networkpolicy", req.NamespacedName)
+			updateFail(r, &ctx, networkPolicy, &err)
+			return ResultRequeue, err
+		} else if isCRInSysNs {
+			err = errors.New("network policy cannot be created in System Namespace")
```

Comment:
Why we cannot create NetworkPolicy in System namespace? For apiserver/coredns service, do we have any NetworkPolicy?